source/Calamari.Shared/Integration/Packages/Download/ArtifactoryPackageDownloader.cs

```diff
@@ -0,0 +1,375 @@
+                    return PackagePhysicalFileMetadata.Build(file, package);
+            }
+
+            return null;
+        }
+
+        public PackagePhysicalFileMetadata DownloadPackage(
+            string packageId,
+            IVersion version,
+            Uri feedUri,
+            ICredentials feedCredentials,
+            string cacheDirectory)
+        {
+            Log.Info("Downloading package {0} v{1} from feed: '{2}'", packageId, version, feedUri);
+            Log.VerboseFormat("Downloaded package will be stored in: '{0}'", cacheDirectory);
+
+            var tempDirectory = fileSystem.CreateTemporaryDirectory();
+            using (new TemporaryDirectory(tempDirectory))
+            {
+                var stagingDir = Path.Combine(tempDirectory, "staging");
+                if (!Directory.Exists(stagingDir))
+                {
+                    Directory.CreateDirectory(stagingDir);
+                }
+                
+                var (downloadUri, fileExtension) = GetDownloadPackagePath(feedUri, packageId, version, feedCredentials);
+                
+                var cachedFileName = PackageName.ToCachedFileName(packageId, version, fileExtension);
+                var downloadPath = Path.Combine(Path.Combine(stagingDir, cachedFileName));
+
+                using (var fileStream = fileSystem.OpenFile(downloadPath, FileAccess.Write))
+                {
+                    using var response = DownloadPackage(downloadUri, feedCredentials);
+                    if (!response.IsSuccessStatusCode)
+                    {
+                        throw new CommandException(
+                                                   $"Failed to download artifact (Status Code {(int)response.StatusCode}). Reason: {response.ReasonPhrase}");
+                    }
+
+#if NET40
+                    response.Content.CopyToAsync(fileStream).Wait();
+#else
+                    response.Content.CopyToAsync(fileStream).GetAwaiter().GetResult();
+#endif
+                }
+
+                var localDownloadName = Path.Combine(cacheDirectory, cachedFileName);
+                fileSystem.MoveFile(downloadPath, localDownloadName);
+
+                return PackagePhysicalFileMetadata.Build(localDownloadName);
+            }
+        }
+
+        private (string downloadUrl, string fileExtension) GetDownloadPackagePath(Uri url, string packageId, IVersion version, ICredentials credentials)
+        {
+            var layoutRegex = variables.Get("ArtifactoryGenericFeed.Regex");
+            var repository = variables.Get("ArtifactoryGenericFeed.Repository");
+
+            var artifactId = packageId.Split('/').Last();
+            var path = packageId.Substring(0, packageId.LastIndexOf('/'));
+            var regex = string.IsNullOrEmpty(layoutRegex) ? new Regex("(?<orgPath>.+?)/(?<module>[^/]+)/(?<module>\\2)-(?<baseRev>[^/]+?)\\.(?<ext>(?:(?!\\d))[^\\-/]+|7z)", RegexOptions.Compiled) : new Regex(layoutRegex, RegexOptions.Compiled);
```

Comment:
I feel like we would always send down the default regex, not sure if we need to hard code it here too